File: 1306-jump-game-iii/1306-jump-game-iii.cpp
class Solution {
public:
    bool canReach(vector<int>& arr, int start) {
        // Rahul Sir
        // bfs
        
        queue<int> q;
        q.push(start);
        
        while(!q.empty()) {
            int idx = q.front();
            q.pop();
            
            if(arr[idx] == 0) return true;
            
            if(arr[idx] < 0) continue;
            
            if(idx + arr[idx] < (int)arr.size()) q.push(idx + arr[idx]);
            if(idx - arr[idx] >= 0) q.push(idx - arr[idx]);

            arr[idx] = -1 * arr[idx];
        }
        return false;
    }
};

File: 1306-jump-game-iii/1306-jump-game-iii.py
class Solution:
    def canReach(self, arr: List[int], start: int) -> bool:
        # Rahul Sir - BFS
        
        queue = deque()
        
        queue.append(start)
        
        while len(queue) > 0:
            idx = queue.popleft()
            
            if arr[idx] == 0: return True
            if arr[idx] < 0: continue
            
            if idx + arr[idx] < len(arr): queue.append(idx + arr[idx])
            if idx - arr[idx] >= 0: queue.append(idx - arr[idx])
            
            arr[idx] *= -1
            
        return False
        